internal/report/snapshot_reporter.go

```diff
@@ -0,0 +1,214 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package report contains reporters for collecting specific information about the collector-
+package report
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/observiq/observiq-otel-collector/internal/report/snapshot"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+// snapShotKind is the kind for the snapshot reporter
+var snapShotKind = "snapshot"
+
+// snapshotConfig specifies what snapshots to collect
+type snapshotConfig struct {
+	// Endpoint is where to send the snapshots
+	Endpoint *endpointConfig `yaml:"endpoint"`
+
+	// Processor is the full ComponentID of the snapshot processor
+	Processor string `yaml:"processor"`
+
+	// PipelineType will be "logs", "metrics", or "traces"
+	PipelineType string `yaml:"pipeline_type"`
+}
+
+// endpointConfig is the configuration of a specific endpoint and full headers to include
+type endpointConfig struct {
+	URL     string      `yaml:"url"`
+	Headers http.Header `yaml:"headers"`
+}
+
+var _ Reporter = (*SnapshotReporter)(nil)
+var _ snapshot.Snapshotter = (*SnapshotReporter)(nil)
+
+// SnapshotReporter tracks and reports snapshots
+type SnapshotReporter struct {
+	client Client
+
+	// minPayloadSize is the minimum number of items to be in a payload
+	minPayloadSize int
+
+	// Buffers
+	logBuffers    map[string]*snapshot.LogBuffer
+	metricBuffers map[string]*snapshot.MetricBuffer
+	traceBuffers  map[string]*snapshot.TraceBuffer
+}
+
+// NewSnapshotReporter creates a new SnapshotReporter with the associated client
+func NewSnapshotReporter(client Client) *SnapshotReporter {
+	return &SnapshotReporter{
+		client:         client,
+		minPayloadSize: 100,
```

Comment:
I notice this isn't configurable; Any worries about that?

My main worry is particularly with metrics, where many data points in a single batch are common. So you might tend to only get datapoints for a single metric source if this is too low.

Reply:
This isn't configurable by design. Since this feature set is targeted at BPOP we chose not to over engineer this. 100 items is good enough to have the feature work in BPOP without having to worry about resizing buffers on the fly.

The first case of the buffer logic replaces the entire buffer with a single payload if it's larger than the `idealSize`. This is likely the most common case at scale. We also have no control on how BPOP chooses to filter down what we send it so the user might be missing out on info even if we send it all just based on how we curate it for them.

Reply:
I think my worry is just that the user will consistently not see metrics from some source, and will take that to mean that the metric is filtered out. It seems like this is fine for a first pass, and might not actually be an issue.